projects/Mallard/src/helpers/queries.ts

```diff
@@ -0,0 +1,249 @@
+            node.query.resolver(
+                node.variables,
+                (resolve as unknown) as LocalResolver,
+                prevValue,
+            ),
+        )
+        node.promise = promise
+        promise
+            .then(
+                value => {
+                    if (node.promise !== promise) return
+                    node.promise = undefined
+                    node.value = value
+                },
+                error => {
+                    if (node.promise !== promise) return
+                    node.promise = undefined
+                    node.error = error
+                },
+            )
+            .then(() => {
+                for (const listener of node.listeners) {
+                    listener()
+                }
+            })
+            .catch(error => {
+                setImmediate(() => {
+                    throw error
+                })
+            })
+
+        const innerTrail = new Set(trail)
+        innerTrail.add(node.key)
+
+        for (const predKey of node.preds) {
+            const pred = this._graph.get(predKey)
+            if (pred === undefined || !pred.deps.has(node.key))
+                throw new Error('inconsistent graph')
+            this._refresh(pred, innerTrail)
+        }
+    }
+
+    private _releaseNode(node: QueryNode): void {
+        if (node.listeners.length > 0 || node.preds.size > 0) return
+        for (const depKey of node.deps) {
+            const dnode = this._graph.get(depKey)
+            if (dnode === undefined) throw new Error('inconsistent state')
+            dnode.preds.delete(node.key)
+            this._releaseNode(dnode)
+        }
+        this._graph.delete(node.key)
+    }
+}
+
+export class Query<Value, Variables> {
+    private static _nextId = 1
+
+    private _resolver: QueryResolver<Value, Variables>
+    private _id: number
+
+    private constructor(resolver: QueryResolver<Value, Variables>) {
+        this._id = Query._nextId++
+        this._resolver = resolver
+    }
+
+    get id() {
+        return this._id
+    }
+
+    get resolver() {
+        return this._resolver
+    }
+
+    static create<Value, Variables>(
+        resolver: QueryResolver<Value, Variables>,
+    ): Query<Value, Variables> {
+        return new Query(resolver)
+    }
+}
+
+export const QueryEnvironmentContext = createContext<
+    QueryEnvironment | undefined
+>(undefined)
+
+export const useQueryEnvironment = (): QueryEnvironment => {
+    const env = useContext(QueryEnvironmentContext)
+    if (env === undefined) throw new Error('no query environment availalbe')
+    return env
+}
+
+export const useQuery = <Value, Variables extends AllowedVariables>(
+    query: Query<Value, Variables>,
+    variables: Variables,
+): QueryResult<Value> => {
+    const env = useQueryEnvironment()
+    const [result, setResult] = useState(env.peek(query, variables))
+    const serializedVars = JSON.stringify(variables)
```

Comment:
Yes I agree we could, though it'd require callers to use `React.useMemo` in a number of cases. On the other hand it'd be generally more performant and less opiniated about `variables`.

Reply:
Wait, no, we can't use ref equality, we do want value equality. The reason why is because if you have two completely unrelated callers (ex. components) providing the same variables (in value), we want to merge the query promises to avoid separate extra resolver calls!

Reply:
For example if I do:

    const result = useQuery(
      ISSUE_QUERY,
      React.useMemo(() => ({issueId}), [issueId]),
    );

somewhere and I do the exact same from a completely different component, we don't want to re-fetch the entire issue a second time. We want to reuse what was cached for `issueId`, even if the components are unrelated. Otherwise it would force these to share the entire variable object all over the place.

Reply:
You are totally right 👌